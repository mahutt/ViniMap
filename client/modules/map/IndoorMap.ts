import { IndoorMap, IndoorMapGeoJSON } from './Types';
import LocalLocations from '@/services/LocalLocations';

import hallGeojson from '@/assets/geojson/hall.json';
import jmsbGeojson from '@/assets/geojson/jmsb.json';
import VLandVEfloors from '@/assets/geojson/VLandVEfloors.json';
import websterGeojson from '@/assets/geojson/webster.json';
import evGeojson from '@/assets/geojson/ev.json';
import erGeojson from '@/assets/geojson/er.json';
import lsGeojson from '@/assets/geojson/ls.json';
import clGeojson from '@/assets/geojson/cl.json';
import fbGeojson from '@/assets/geojson/fb.json';
import nunsGeojson from '@/assets/geojson/nuns.json';
import vaGeojson from '@/assets/geojson/va.json';
import sdGeojson from '@/assets/geojson/sd.json';
import pcGeojson from '@/assets/geojson/pc.json';
import raGeojson from '@/assets/geojson/ra.json';
import psGeojson from '@/assets/geojson/ps.json';
import pyGeojson from '@/assets/geojson/py.json';
import rfGeojson from '@/assets/geojson/rf.json';
import ccGeojson from '@/assets/geojson/cc.json';
import adGeojson from '@/assets/geojson/ad.json';
import fcGeojson from '@/assets/geojson/fc.json';
import cjGeojson from '@/assets/geojson/cj.json';
import siGeojson from '@/assets/geojson/si.json';
import hcGeojson from '@/assets/geojson/hc.json';
import hbGeojson from '@/assets/geojson/hb.json';
import haGeojson from '@/assets/geojson/ha.json';
import huGeojson from '@/assets/geojson/hu.json';
import spGeojson from '@/assets/geojson/sp.json';
import geGeojson from '@/assets/geojson/ge.json';

import GeojsonService from '@/services/GeojsonService';
import type { Feature, Polygon } from 'geojson';
import { center } from '@turf/turf';

const getAllRooms = (geojson: IndoorMapGeoJSON): Feature<Polygon>[] => {
  const rooms: Feature<Polygon>[] = [];
  geojson.features.forEach((feature) => {
    if (feature.properties?.ref) {
      rooms.push(feature as Feature<Polygon>);
    }
  });
  return rooms;
};

const rawIndoorMaps = [
  {
    id: 'Hall Building (H)',
    mapboxIds: [22080570],
    geojson: hallGeojson,
    addr: '1455 Blvd. De Maisonneuve Ouest, Montreal, Quebec H3G 1M8',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Webster Library (LB)',
    mapboxIds: [
      1110900192, 1110900193, 1110900197, 1110900196, 1110900194, 1110900195, 1110900190,
      1110900191,
    ],
    geojson: websterGeojson,
    addr: 'Pavillion J.W. McConnell Bldg, 1400 Maisonneuve Blvd W, Montreal, Quebec H3G 1M8',
    opening_hours: '24/7',
  },
  {
    id: 'Engineering, CS, and VA Integrated Complex (EV)',
    mapboxIds: [1110145740, 1110145744, 1110145741, 356036803, 1110145743],
    geojson: evGeojson,
    addr: '1515 Saint-Catherine St W, Montreal, Quebec H3G 2H7',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'ER Building (ER)',
    mapboxIds: [103896385],
    geojson: erGeojson,
    addr: '2155 Guy St, Montreal, Quebec H3H 2L9',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Learning Square (LS)',
    mapboxIds: [979438074],
    geojson: lsGeojson,
    addr: '1535 Blvd. De Maisonneuve Ouest, Montreal, Quebec H3G 1M9',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'John Molson Building (MB)',
    mapboxIds: [1109876433, 1109876436, 1109876434, 1109876435],
    geojson: jmsbGeojson,
    addr: '1600 Blvd. De Maisonneuve Ouest, Montreal, Quebec H3H 0A1',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Faubourg Tower (FB)',
    mapboxIds: [103248058],
    geojson: fbGeojson,
    addr: '1600 Saint-Catherine St W, Montreal, Quebec H3H 2S7',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Grey Nuns Building (GN)',
    mapboxIds: [17189298],
    geojson: nunsGeojson,
    addr: '1190 Guy St, Montreal, Quebec H3H 2L4',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'CL Building (CL)',
    mapboxIds: [103248055],
    geojson: clGeojson,
    addr: '1665 Saint-Catherine St W, Montreal, Quebec H3H 2S7',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Visual Arts Building (VA)',
    mapboxIds: [103521746],
    geojson: vaGeojson,
    addr: '1395 René-Lévesque Blvd W, Montreal, Quebec H3G 2M5',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Vanier Library (VL / VE)',
    mapboxIds: [604324442, 604324441, 47332009, 604324443],
    geojson: VLandVEfloors,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: '24/7',
  },
  {
    id: 'Stinger Dome (SD)',
    mapboxIds: [129437736],
    geojson: sdGeojson,
    addr: '7200 Sherbrooke St W, Montreal, Quebec H4B 1R3',
    opening_hours: 'Mo-Su 09:00-22:00',
  },
  {
    id: 'Perform Center (PC)',
    mapboxIds: [604019520],
    geojson: pcGeojson,
    addr: '7200 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 06:30-22:00; Sa-Su 08:00-18:00',
  },
  {
    id: 'Recreation and Althletics Complex (RA)',
    mapboxIds: [47332005],
    geojson: raGeojson,
    addr: '7200 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 06:30-22:00; Sa-Su 08:00-18:00',
  },
  {
    id: 'Physical Services (PS)',
    mapboxIds: [47332006],
    geojson: psGeojson,
    addr: 'Montreal, Quebec H4B 2B9',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Psychology Building (PY)',
    mapboxIds: [17887805],
    geojson: pyGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Loyola Jesuit Hall and Conference Center (RF)',
    mapboxIds: [604324455],
    geojson: rfGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1E1',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Central Building (CC)',
    mapboxIds: [604324456],
    geojson: ccGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1E1',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Administration Building (AD)',
    mapboxIds: [604324457],
    geojson: adGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1E1',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'F.C. Smith Building (FC)',
    mapboxIds: [47332003],
    geojson: fcGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Communication Studies and Journalism Building (CJ)',
    mapboxIds: [47332007],
    geojson: cjGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 08:00-18:00',
  },
  {
    id: 'Saint Ignatius of Loyola Church (SI)',
    mapboxIds: [47331997],
    geojson: siGeojson,
    addr: '4455 Rue West Broadway, Montréal, QC H4B 2A7',
    opening_hours: 'Mo-Fr 08:00-16:00; Sa 15:45-17:30; Su 10:00-11:00',
  },
  {
    id: 'Hingston Hall (HC)',
    mapboxIds: [604324438],
    geojson: hcGeojson,
    addr: 'Hingston Hall C, Montreal, Quebec H4B, Canada',
    opening_hours: '24/7',
  },
  {
    id: 'Hingston Hall (HB)',
    mapboxIds: [17887804],
    geojson: hbGeojson,
    addr: 'Hingston Hall B, Montreal, Quebec H4B, Canada',
    opening_hours: '24/7',
  },
  {
    id: 'Hingston Hall (HA)',
    mapboxIds: [17887803],
    geojson: haGeojson,
    addr: 'Hingston Hall A, Montreal, Quebec H4B, Canada',
    opening_hours: '24/7',
  },
  {
    id: 'Applied Science Hub (HU)',
    mapboxIds: [795012497],
    geojson: huGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Richard J. Renaud Science Complex (SP)',
    mapboxIds: [47331993],
    geojson: spGeojson,
    addr: '3475 Rue West Broadway, Montréal, QC H4B 2A7',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
  {
    id: 'Centre for Structural and Functional Genomics (GE)',
    mapboxIds: [545014554],
    geojson: geGeojson,
    addr: '7141 Sherbrooke St W, Montreal, Quebec H4B 1R6',
    opening_hours: 'Mo-Fr 09:00-17:00',
  },
];

const indoorMaps: IndoorMap[] = rawIndoorMaps.map((rawIndoorMap) => {
  const { bounds, levelsRange } = GeojsonService.extractLevelsRangeAndBounds(
    rawIndoorMap.geojson as IndoorMapGeoJSON
  );
  const indoorMap: IndoorMap = {
    id: rawIndoorMap.id,
    mapboxIds: rawIndoorMap.mapboxIds,
    bounds: bounds,
    geojson: rawIndoorMap.geojson as IndoorMapGeoJSON,
    levelsRange: levelsRange,
    addr: rawIndoorMap.addr,
    opening_hours: rawIndoorMap.opening_hours,
  };
  const rooms = getAllRooms(indoorMap.geojson);
  for (const room of rooms) {
    const roomRef = room.properties?.ref;
    if (typeof roomRef !== 'string') {
      continue;
    }
    LocalLocations.getInstance().add(roomRef, (name: string) => {
      return {
        name,
        coordinates: center(room).geometry.coordinates,
        data: {
          address: indoorMap.id,
          level: parseFloat(room.properties?.level),
          indoorMap: indoorMap,
          ref: roomRef,
          feature: room,
        },
      };
    });
  }
  return indoorMap;
});

const campusMapboxIds = indoorMaps.flatMap((indoorMap) => indoorMap.mapboxIds);

export { indoorMaps, campusMapboxIds };
