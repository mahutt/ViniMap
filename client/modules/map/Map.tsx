import { StyleSheet } from 'react-native';
import Mapbox from '@rnmapbox/maps';
import React from 'react';
Mapbox.setAccessToken(process.env.EXPO_PUBLIC_MAPBOX_ACCESS_TOKEN as string);

// downtown concordia campus (sgw)
const DEFAULT_COORDINATES = {
  longitude: -73.5789,
  latitude: 45.4973,
};

export default function Map() {
  const mapRef = React.useRef<Mapbox.MapView | null>(null);
  const cameraRef = React.useRef<Mapbox.Camera | null>(null);
  return (
    <Mapbox.MapView ref={mapRef} style={styles.map} styleURL={Mapbox.StyleURL.Street}>
      <Mapbox.Camera
        ref={cameraRef}
        zoomLevel={15}
        centerCoordinate={[DEFAULT_COORDINATES.longitude, DEFAULT_COORDINATES.latitude]}
        animationMode="flyTo"
        animationDuration={2000}
      />
    </Mapbox.MapView>
  );
}

const styles = StyleSheet.create({
  map: {
    flex: 1,
    width: '100%',
    height: '100%',
  },
});
