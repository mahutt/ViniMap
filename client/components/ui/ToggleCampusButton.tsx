import React, { useState } from "react";
import { View, Text, TouchableOpacity, StyleSheet, StyleProp, ViewStyle } from "react-native";
import { useMap } from "@/modules/map/MapContext"; 

interface ToggleCampusButtonProps {
  style?: StyleProp<ViewStyle>; 
}

const ToggleCampusButton: React.FC<ToggleCampusButtonProps> = ({ style }) => {
  const [selectedLocation, setSelectedLocation] = useState<"SGW" | "LOY">("SGW");
  const { flyTo } = useMap();

  const SGW_COORDINATES: [number, number] = [-73.5789, 45.4973];
  const LOY_COORDINATES: [number, number] = [-73.6391, 45.4581];

  const handleToggle = (location: "SGW" | "LOY") => {
    setSelectedLocation(location);
    if (location === "SGW") {
      flyTo(SGW_COORDINATES);
    } else {
      flyTo(LOY_COORDINATES);
    }
  };

  return (
    <View style={[styles.container, styles.toggleButton, style]}>
      <TouchableOpacity
        style={[
          styles.button,
          selectedLocation === "SGW" ? styles.activeButton : null,
        ]}
        onPress={() => handleToggle("SGW")}
      >
        <Text
          style={[
            styles.text,
            selectedLocation === "SGW" ? styles.activeText : null,
          ]}
        >
          SGW
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        style={[
          styles.button,
          selectedLocation === "LOY" ? styles.activeButton : null,
        ]}
        onPress={() => handleToggle("LOY")}
      >
        <Text
          style={[
            styles.text,
            selectedLocation === "LOY" ? styles.activeText : null,
          ]}
        >
          LOY
        </Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#f3f3f3",
    borderRadius: 25,
    padding: 5,
    shadowColor: "#000",
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.2,
    shadowRadius: 4,
    elevation: 5,
    width: 138,
    position: "absolute",
    borderWidth: 2,
    borderColor: "#800000",
    zIndex: 2,
  },
  button: {
    flex: 1,
    paddingVertical: 2,
    paddingHorizontal: 16,
    alignItems: "center",
    justifyContent: "center",
    borderRadius: 25,
  },
  activeButton: {
    backgroundColor: "white",
  },
  text: {
    fontSize: 12,
    fontWeight: "bold",
    color: "#800000",
  },
  activeText: {
    color: "#800000",
  },
  toggleButton: {
    top: 125, 
    left: 39, 
    right: 20, 
  },
});

export default ToggleCampusButton;
