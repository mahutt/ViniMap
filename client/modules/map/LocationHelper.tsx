import { Location } from './MapContext';
import CoordinateService from '@/Services/CoordinateService';

export const getCurrentLocationAsStart = async (setLocation: (location: Location) => void) => {
  try {
    console.log('getCurrentLocationAsStart: Attempting to get coordinates');
    const tempCoordinates = await CoordinateService.getCurrentCoordinates();

    console.log('Coordinates retrieved:', tempCoordinates);

    if (tempCoordinates) {
      console.log('Setting location with coordinates:', tempCoordinates);
      setLocation({
        name: 'Current location',
        coordinates: tempCoordinates,
      });
      console.log('Location set successfully');
    } else {
      console.log('No coordinates available');
    }
  } catch (error) {
    console.error('Detailed error in getCurrentLocationAsStart:', error);
  }
};
