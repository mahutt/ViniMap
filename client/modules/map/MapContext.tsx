import React, { createContext, useContext, useRef, useState, useMemo } from 'react';
import Mapbox from '@rnmapbox/maps';
import { getRoute } from './MapService';
import { Coordinates } from './Types';

export interface Location {
  name: string | null;
  coordinates: Coordinates;
  data?: any;
}

export enum MapState {
  Idle,
  RoutePlanning,
  Information,
  SelectingStartLocation,
  SelectingEndLocation,
}

const DEFAULT_COORDINATES: Coordinates = [-73.5789, 45.4973];

type MapContextType = {
  mapRef: React.RefObject<Mapbox.MapView>;
  cameraRef: React.RefObject<Mapbox.Camera>;
  centerCoordinate: [number, number];
  zoomLevel: number;

  pitchLevel: number;
  setPitchLevel: (pitchLevel: number) => void;

  state: MapState;
  startLocation: Location | null;
  endLocation: Location | null;
  routeCoordinates: Coordinates[];
  setCenterCoordinate: (centerCoordinate: [number, number]) => void;
  setZoomLevel: (zoomLevel: number) => void;
  setState: (state: MapState) => void;
  setStartLocation: (startLocation: Location | null) => void;
  setEndLocation: (endLocation: Location | null) => void;
  flyTo: (coords: [number, number], zoomLevel?: number) => void;
  loadRouteFromCoordinates: (
    startCoordinates: Coordinates,
    endCoordinates: Coordinates,
    mode?: string
  ) => Promise<void>;
};

const MapContext = createContext<MapContextType | undefined>(undefined);

export const MapProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const mapRef = useRef<Mapbox.MapView | null>(null);
  const cameraRef = useRef<Mapbox.Camera | null>(null);
  const [centerCoordinate, setCenterCoordinate] = useState<[number, number]>(DEFAULT_COORDINATES);
  const [zoomLevel, setZoomLevel] = useState(15);
  const [pitchLevel, setPitchLevel] = useState(0);
  const [state, setState] = useState<MapState>(MapState.Idle);

  const setStateWithLogging = (newState: MapState) => {
    switch (newState) {
      case MapState.SelectingStartLocation:
        setStartLocation(null);
        break;
      case MapState.RoutePlanning:
        break;
    }

    setState(newState);
  };

  const [startLocation, setStartLocation] = useState<Location | null>(null);
  const [endLocation, setEndLocation] = useState<Location | null>(null);
  const [routeCoordinates, setRouteCoordinates] = useState<Coordinates[]>([]);

  const flyTo = useMemo(
    () => (newCenterCoordinate: [number, number], newZoomLevel?: number) => {
      if (cameraRef.current) {
        cameraRef.current.setCamera({
          centerCoordinate: newCenterCoordinate,
          zoomLevel: newZoomLevel ?? zoomLevel,
          animationDuration: 2000,
          pitch: pitchLevel,
        });
        setCenterCoordinate(newCenterCoordinate);
        if (newZoomLevel) setZoomLevel(newZoomLevel);
      }
    },
    [zoomLevel, pitchLevel]
  );

  const loadRouteFromCoordinates = async (
    startCoordinates: Coordinates,
    endCoordinates: Coordinates,
    mode = 'walking'
  ): Promise<void> => {
    if (!startCoordinates || !endCoordinates) {
      return;
    }
    return getRoute(startCoordinates, endCoordinates, mode)
      .then((data) => {
        if (data?.coordinates) {
          if (data.coordinates.length > 0) {
            flyTo(data.coordinates[0], zoomLevel);
          }
          setRouteCoordinates(data.coordinates);
        }
      })
      .catch((error) => {
        console.error('Error setting route:', error);
      });
  };

  const value = useMemo(
    () => ({
      mapRef,
      cameraRef,
      centerCoordinate,
      zoomLevel,
      pitchLevel,
      setPitchLevel,
      state,
      startLocation,
      endLocation,
      routeCoordinates,
      setCenterCoordinate,
      setZoomLevel,
      setState: setStateWithLogging,
      setStartLocation,
      setEndLocation,
      flyTo,
      loadRouteFromCoordinates,
    }),
    [
      centerCoordinate,
      zoomLevel,
      state,
      startLocation,
      endLocation,
      routeCoordinates,
      flyTo,
      pitchLevel,
    ]
  );

  return <MapContext.Provider value={value}>{children}</MapContext.Provider>;
};

export const useMap = () => {
  const context = useContext(MapContext);
  if (context === undefined) {
    throw new Error('useMap must be used within a MapProvider');
  }
  return context;
};
export { Coordinates };
