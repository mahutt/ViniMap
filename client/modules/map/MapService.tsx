import ShuttleCalculatorService from '@/services/ShuttleCalculatorService';
import { Coordinates, Location } from './MapContext';
import { calculateEuclideanDistance } from './MapUtils';
import { Route } from './Types';

const MAPBOX_ACCESS_TOKEN = process.env.EXPO_PUBLIC_MAPBOX_ACCESS_TOKEN as string;
let GOOGLE_API_KEY = process.env.EXPO_PUBLIC_GOOGLEMAPS_API_KEY as string;

const PROXIMITY_COORDINTATES = {
  longitude: -73.57791396549962, // Concordia SGW Campus Longitude
  latitude: 45.495102086770814, // Concordia SGW Campus  Latitude
};

const getLocations = async (locationQuery: string): Promise<Location[]> => {
  const response = await fetch(
    `https://api.mapbox.com/search/geocode/v6/forward?q=${encodeURIComponent(
      locationQuery
    )}&proximity=${PROXIMITY_COORDINTATES.longitude},${
      PROXIMITY_COORDINTATES.latitude
    }&access_token=${MAPBOX_ACCESS_TOKEN}`
  );

  const data = await response.json();
  if (data?.features) {
    return data.features.map((feature: any) => ({
      name: `${feature.properties.name}, ${feature.properties.place_formatted ?? ''}`,
      coordinates: feature.geometry.coordinates as Coordinates,
    }));
  }
  return [];
};

const getRoute = async (
  startCoordinates: Coordinates,
  endCoordinates: Coordinates,
  mode: string
): Promise<Route | null> => {
  if (mode === 'shuttle') {
    return getRouteForShuttle(startCoordinates, endCoordinates);
  }

  const url = `https://api.mapbox.com/directions/v5/mapbox/${mode}/${startCoordinates[0]},${startCoordinates[1]};${endCoordinates[0]},${endCoordinates[1]}?alternatives=false&annotations=duration,distance&continue_straight=true&geometries=geojson&language=en&overview=full&steps=true&access_token=${MAPBOX_ACCESS_TOKEN}`;
  const response = await fetch(url);
  const data = await response.json();
  if (data?.routes[0]) {
    const route = data.routes[0];
    return {
      duration: route.duration,
      distance: route.distance,
      segments: [
        {
          id: mode,
          type: 'solid',
          steps: route.geometry.coordinates as Coordinates[],
        },
      ],
    };
  }
  return null;
};

const fetchLocationData = async (coordinates: Coordinates) => {
  const radius = 50;
  const apiUrl = `https://maps.googleapis.com/maps/api/place/nearbysearch/json?location=${coordinates[1]},${coordinates[0]}&radius=${radius}&key=${GOOGLE_API_KEY}`;
  try {
    const response = await fetch(apiUrl);
    const data = await response.json();

    if (data.results.length > 0) {
      return {
        address: data.results[0]?.name || 'Address not available',
        name: data.results[1]?.name || 'Name not available',
        isOpen: Boolean(data.results[1].opening_hours),
      };
    }
  } catch (error) {
    console.error('Error fetching data:', error);
  }
};

const getRouteForShuttle = async (
  startCoordinates: Coordinates,
  endCoordinates: Coordinates
): Promise<Route | null> => {
  const MAX_EUCLIDEAN_DISTANCE = 0.01508;

  const SgwCoords: Coordinates = [-73.5784711, 45.4970661];
  const LoyolaCoords: Coordinates = [-73.6393324, 45.4577857];

  let startBusStop: Coordinates = [0, 0];
  let endBusStop: Coordinates = [0, 0];

  let isStartSgw = false;
  let isStartLoyola = false;
  let isEndSgw = false;
  let isEndLoyola = false;

  const distanceToShuttleSgwFromStart = calculateEuclideanDistance(SgwCoords, startCoordinates);
  const distanceToShuttleLoyolaFromStart = calculateEuclideanDistance(
    LoyolaCoords,
    startCoordinates
  );
  const distanceToShuttleSgwFromEnd = calculateEuclideanDistance(SgwCoords, endCoordinates);
  const distanceToShuttleLoyolaFromEnd = calculateEuclideanDistance(LoyolaCoords, endCoordinates);

  if (distanceToShuttleSgwFromStart <= MAX_EUCLIDEAN_DISTANCE) {
    isStartSgw = true;
    startBusStop = SgwCoords;
  }
  if (distanceToShuttleLoyolaFromStart <= MAX_EUCLIDEAN_DISTANCE) {
    isStartLoyola = true;
    startBusStop = LoyolaCoords;
  }
  if (distanceToShuttleSgwFromEnd <= MAX_EUCLIDEAN_DISTANCE) {
    isEndSgw = true;
    endBusStop = SgwCoords;
  }
  if (distanceToShuttleLoyolaFromEnd <= MAX_EUCLIDEAN_DISTANCE) {
    isEndLoyola = true;
    endBusStop = LoyolaCoords;
  }

  if (!((isStartLoyola && isEndSgw) || (isStartSgw && isEndLoyola))) {
    return null;
  }

  const start_Walk_ShuttleStart = await getRoute(startCoordinates, startBusStop, 'walking');
  const shuttleStart_Drive_shuttleEnd = await getRoute(startBusStop, endBusStop, 'driving');
  const shuttleEnd_Walk_end = await getRoute(endBusStop, endCoordinates, 'walking');

  if (!start_Walk_ShuttleStart || !shuttleStart_Drive_shuttleEnd || !shuttleEnd_Walk_end) {
    return null;
  }

  const firstWalkCoords = start_Walk_ShuttleStart?.segments[0].steps ?? [];
  const shuttleCoordinates = shuttleStart_Drive_shuttleEnd?.segments[0].steps ?? [];
  const secondWalkCoordinates = shuttleEnd_Walk_end?.segments[0].steps ?? [];

  const daysMap: { [key: number]: string } = {
    1: 'Monday-Thursday',
    2: 'Monday-Thursday',
    3: 'Monday-Thursday',
    4: 'Monday-Thursday',
    5: 'Friday',
  };
  const today = new Date().getDay();
  const dayKey = daysMap[today] ?? 'Monday-Thursday';
  const currentTime = ShuttleCalculatorService.getCurrentTime();

  const departureCampus: 'LOY' | 'SGW' = isStartSgw ? 'SGW' : 'LOY';
  const nextShuttleDuration = ShuttleCalculatorService.getNextDepartureTime(
    dayKey,
    currentTime,
    departureCampus
  );

  const shuttleDurationInSeconds =
    Number(nextShuttleDuration.replace('m', '').replace('h', '')) *
    (nextShuttleDuration.includes('h') ? 3600 : 60);

  if (isNaN(Number(shuttleDurationInSeconds))) {
    return null;
  }

  const finalDuration =
    Number(start_Walk_ShuttleStart.duration) +
    Number(shuttleStart_Drive_shuttleEnd.duration) +
    Number(shuttleEnd_Walk_end.duration) +
    shuttleDurationInSeconds;

  const finalDistance =
    Number(start_Walk_ShuttleStart.distance) +
    Number(shuttleStart_Drive_shuttleEnd.distance) +
    Number(shuttleEnd_Walk_end.distance);

  return {
    duration: finalDuration,
    distance: finalDistance,
    segments: [
      {
        id: 'firstWalk',
        type: 'dashed',
        steps: firstWalkCoords,
      },
      {
        id: 'shuttle',
        type: 'solid',
        steps: shuttleCoordinates,
      },
      {
        id: 'secondWalk',
        type: 'dashed',
        steps: secondWalkCoordinates,
      },
    ],
  };
};

const formatDuration = (seconds: number | null): string => {
  if (seconds === null) return 'Unavailable';
  const minutes = Math.round(seconds / 60);
  if (minutes >= 60) {
    const hours = Math.floor(minutes / 60);
    const remainingMinutes = minutes % 60;
    return `${hours}h ${remainingMinutes}m`;
  }
  return `${minutes} min`;
};

export { getLocations, getRoute, fetchLocationData, formatDuration };
