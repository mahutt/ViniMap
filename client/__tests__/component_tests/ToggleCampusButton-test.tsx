import { render, fireEvent } from '@testing-library/react-native';
import ToggleCampusButton from '@/components/ui/ToggleCampusButton';
import { MapProvider } from '@/modules/map/MapContext';
import React from 'react';

const mockFlyTo = jest.fn();

jest.mock('@/modules/map/MapContext', () => {
  const actual = jest.requireActual('@/modules/map/MapContext');

  const mockUseMap = () => ({
    flyTo: mockFlyTo,
  });

  return {
    ...actual,
    MapProvider: actual.MapProvider,
    useMap: mockUseMap,
  };
});

const isTextActive = (textElement: any): boolean => {
  if (!textElement?.props?.style) {
    return false;
  }

  const styles = textElement.props.style;

  if (Array.isArray(styles)) {
    return styles.some((style: any) => style && style.color === 'white');
  } else if (typeof styles === 'object') {
    return styles.color === 'white';
  }

  return false;
};

describe('<ToggleCampusButton />', () => {
  beforeEach(() => {
    mockFlyTo.mockClear();
  });

  test('ToggleCampusButton renders correctly', () => {
    const tree = render(
      <MapProvider>
        <ToggleCampusButton />
      </MapProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Default selected location is SGW', () => {
    const { getByText } = render(
      <MapProvider>
        <ToggleCampusButton />
      </MapProvider>
    );

    const sgwText = getByText('SGW');
    const loyText = getByText('LOY');

    expect(isTextActive(sgwText)).toBe(true);
    expect(isTextActive(loyText)).toBe(false);
  });

  test('Clicking LOY button calls flyTo with LOY coordinates', () => {
    const { getByText } = render(
      <MapProvider>
        <ToggleCampusButton />
      </MapProvider>
    );

    fireEvent.press(getByText('LOY'));

    expect(mockFlyTo).toHaveBeenCalledWith([-73.6391, 45.4581]);

    const loyText = getByText('LOY');
    const sgwText = getByText('SGW');

    expect(isTextActive(loyText)).toBe(true);
    expect(isTextActive(sgwText)).toBe(false);
  });

  test('Clicking SGW button after LOY calls flyTo with SGW coordinates', () => {
    const { getByText } = render(
      <MapProvider>
        <ToggleCampusButton />
      </MapProvider>
    );

    fireEvent.press(getByText('LOY'));

    fireEvent.press(getByText('SGW'));

    expect(mockFlyTo).toHaveBeenNthCalledWith(2, [-73.5789, 45.4973]);

    const sgwText = getByText('SGW');
    const loyText = getByText('LOY');

    expect(isTextActive(sgwText)).toBe(true);
    expect(isTextActive(loyText)).toBe(false);
  });
});
