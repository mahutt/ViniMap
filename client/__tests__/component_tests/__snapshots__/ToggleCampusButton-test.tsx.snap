// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleCampusButton /> ToggleCampusButton renders correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#f3f3f3",
        "borderColor": "#800000",
        "borderRadius": 25,
        "borderWidth": 2,
        "elevation": 5,
        "flexDirection": "row",
        "height": 30,
        "justifyContent": "center",
        "padding": 3,
        "position": "absolute",
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 4,
        "width": 138,
        "zIndex": 2,
      },
      {
        "left": 39,
        "right": 20,
        "top": 125,
      },
      undefined,
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderRadius": 25,
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 16,
        "paddingVertical": 3,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#800000",
            "fontSize": 10,
            "fontWeight": "bold",
          },
          null,
        ]
      }
    >
      LOY
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#800000",
        "borderRadius": 25,
        "flex": 1,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 16,
        "paddingVertical": 3,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#800000",
            "fontSize": 10,
            "fontWeight": "bold",
          },
          {
            "color": "white",
          },
        ]
      }
    >
      SGW
    </Text>
  </View>
</View>
`;
